src/System.Management.Automation/engine/CommandDiscovery.cs

```diff
@@ -409,17 +408,69 @@ private static void VerifyRequiredSnapins(IEnumerable<PSSnapInSpecification> req
             }
         }
 
-        // This method verifies the following 3 elements of #Requires statement
+        // This method verifies the following 5 elements of #Requires statement
         // #Requires -RunAsAdministrator
         // #Requires -PSVersion
+        // #Requires -MaximumPSVersion
         // #Requires -PSEdition
         // #Requires -Module
+        // #Requires -OS
         internal static void VerifyScriptRequirements(ExternalScriptInfo scriptInfo, ExecutionContext context)
         {
             VerifyElevatedPrivileges(scriptInfo);
             VerifyPSVersion(scriptInfo);
+            VerifyMaximumPSVersion(scriptInfo);
             VerifyPSEdition(scriptInfo);
             VerifyRequiredModules(scriptInfo, context);
+            VerifyOSTypes(scriptInfo);
+        }
+
+        internal static void VerifyMaximumPSVersion(ExternalScriptInfo scriptInfo)
+        {
+            Version requiresMaximumPSVersion = scriptInfo.RequiresMaximumPSVersion;
+            // check if a maximum PS version is specified
+            if (requiresMaximumPSVersion == null)
+            {
+                return;
+            }
+            // check if the maximum version is a legit PS version.
+            if (!Utils.IsPSVersionSupported(requiresMaximumPSVersion))
+            {
+                ScriptRequiresException scriptRequiresException =
+                    new ScriptRequiresException(
+                        scriptInfo.Name,
+                        requiresMaximumPSVersion,
+                        PSVersionInfo.PSVersion.ToString(),
+                        "ScriptRequiresUnmatchedPSVersion");
+                throw scriptRequiresException;
```

Comment:
Seems like it's simpler to just throw directly and not define the local var here?

```csharp
                throw new ScriptRequiresException(
                    scriptInfo.Name,
                    requiresMaximumPSVersion,
                    PSVersionInfo.PSVersion.ToString(),
                    "ScriptRequiresUnmatchedPSVersion");
```

Reply:
I agree, even if the existing style does it this way.

Reply:
Agreed, even though it's the existing style